src/main/java/hudson/security/UserAttributesHelper.java

```diff
@@ -0,0 +1,190 @@
+    // https://support.microsoft.com/en-us/help/305144/how-to-use-the-useraccountcontrol-flags-to-manipulate-user-account-pro
+    private static String ATTR_USER_ACCOUNT_CONTROL = "userAccountControl";
+    private static String ATTR_ACCOUNT_EXPIRES = "accountExpires";
+    // for Windows Server 2003-based domain
+    private static String ATTR_USER_ACCOUNT_CONTROL_COMPUTED = "msDS-User-Account-Control-Computed";
+    // for ADAM (Active Directory Application Mode), replace the ADS_UF_DISABLED
+    private static String ATTR_USER_ACCOUNT_DISABLED = "msDS-UserAccountDisabled";
+    // for ADAM, replace the ADS_UF_PASSWORD_EXPIRED
+    private static String ATTR_USER_PASSWORD_EXPIRED = "msDS-UserPasswordExpired";
+
+    // https://docs.microsoft.com/en-us/windows/desktop/adschema/a-accountexpires
+    // constant names follow the code in Iads.h
+    private static long ACCOUNT_NO_EXPIRATION = 0x7FFFFFFFFFFFFFFFL;
+    private static int ADS_UF_DISABLED = 0x0002;
+    private static int ADS_UF_LOCK_OUT = 0x0010;
+    private static int ADS_UF_PASSWORD_EXPIRED = 0x800000;
+
+    public static boolean checkIfUserIsEnabled(@Nonnull Attributes user) {
+        try {
+            String userAccountControl = getStringAttribute(user, ATTR_USER_ACCOUNT_CONTROL);
+            if (userAccountControl != null) {
+                int uacAsInt = Integer.parseInt(userAccountControl);
+                if ((uacAsInt & ADS_UF_DISABLED) == ADS_UF_DISABLED) {
+                    return false;
+                }
+            }
+
+            String adamUserAccountDisabled = getStringAttribute(user, ATTR_USER_ACCOUNT_DISABLED);
+            if (adamUserAccountDisabled != null) {
+                if (adamUserAccountDisabled.equals("true")) {
+                    return false;
+                } else {
+                    return true;
+                }
+            }
+
+            return true;
+        } catch (NamingException e) {
+            return true;
+        }
+    }
+
+    public static boolean checkIfAccountNonExpired(@Nonnull Attributes user) {
+        try {
+            String accountExpirationDate = getStringAttribute(user, ATTR_ACCOUNT_EXPIRES);
+            if (accountExpirationDate != null) {
+                long expirationAsLong = Long.parseLong(accountExpirationDate);
+                if (expirationAsLong == 0L || expirationAsLong == ACCOUNT_NO_EXPIRATION) {
+                    return true;
+                }
+
+                long nowIn100NsFromJan1601 = getNowIn100NsFromJan1601();
+                boolean expired = expirationAsLong < nowIn100NsFromJan1601;
+                return !expired;
+            }
+
+            return true;
+        } catch (NamingException e) {
+            return true;
+        }
+    }
+
+    // documentation: https://docs.microsoft.com/en-us/windows/desktop/adschema/a-accountexpires
+    // code inspired by https://community.oracle.com/thread/1157460
+    private static long getNowIn100NsFromJan1601() {
```

Comment:
How about a name like `getWin32EpochHundredNanos` or something? Key changes being remove the `100Ns` and spell it out somehow, and the `Jan1601` is a little awkward in my opinion. Your variable names within the function are more descriptive.